### PR TITLE
SharedBuffer IPC encoders have duplicated and redundant code

### PR DESCRIPTION
**Pull request by jyavenard**
#### cea1d90e9b0fb95ff4489c9badea98404714ec58
<pre>
SharedBuffer IPC encoders have duplicated and redundant code
<a href="https://bugs.webkit.org/show_bug.cgi?id=241407">https://bugs.webkit.org/show_bug.cgi?id=241407</a>
&lt;rdar://94621576 &gt;

Reviewed by Kimmo Kinnunen.

Bug 240160 allowed to greatly simplify the serialisation of SharedBuffer objects,
so make use of it.

* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder&lt;DOMCacheEngine::Record &gt;::encode):
(IPC::ArgumentCoder&lt;DOMCacheEngine::Record &gt;::decode):
(IPC::ArgumentCoder&lt;PasteboardCustomData::Entry &gt;::encode):
(IPC::ArgumentCoder&lt;PasteboardCustomData::Entry &gt;::decode):
(IPC::ArgumentCoder&lt;PasteboardWebContent &gt;::encode):
(IPC::ArgumentCoder&lt;PasteboardWebContent &gt;::decode):
(IPC::ArgumentCoder&lt;PasteboardImage &gt;::encode):
(IPC::ArgumentCoder&lt;PasteboardImage &gt;::decode):
(IPC::ArgumentCoder&lt;RawFile &gt;::encode):
(IPC::ArgumentCoder&lt;RawFile &gt;::decode):
(IPC::ArgumentCoder&lt;PromisedAttachmentInfo &gt;::encode):
(IPC::ArgumentCoder&lt;PromisedAttachmentInfo &gt;::decode):
(IPC::ArgumentCoder&lt;WebCore::FragmentedSharedBuffer &gt;::encode):
(IPC::ArgumentCoder&lt;WebCore::FragmentedSharedBuffer &gt;::decode):
(IPC::ArgumentCoder&lt;WebCore::SharedBuffer &gt;::encode):
(IPC::ArgumentCoder&lt;WebCore::SharedBuffer &gt;::decode):
(IPC::ArgumentCoder&lt;WebCore::ScriptBuffer &gt;::encode):
(IPC::ArgumentCoder&lt;WebCore::ScriptBuffer &gt;::decode):
(IPC::ArgumentCoder&lt;WebCore::CDMInstanceSession::Message &gt;::encode):
(IPC::ArgumentCoder&lt;WebCore::CDMInstanceSession::Message &gt;::decode):
(IPC::encodeSharedBuffer): Deleted.
(IPC::decodeSharedBuffer): Deleted.
(IPC::encodeTypesAndData): Deleted.
(IPC::decodeTypesAndData): Deleted.
(IPC::ArgumentCoder&lt;RefPtr&lt;WebCore::FragmentedSharedBuffer &gt; &gt;::encode): Deleted.
(IPC::ArgumentCoder&lt;RefPtr&lt;WebCore::FragmentedSharedBuffer &gt; &gt;::decode): Deleted.
(IPC::ArgumentCoder&lt;Ref&lt;WebCore::FragmentedSharedBuffer &gt; &gt;::encode): Deleted.
(IPC::ArgumentCoder&lt;Ref&lt;WebCore::FragmentedSharedBuffer &gt; &gt;::decode): Deleted.
(IPC::ArgumentCoder&lt;RefPtr&lt;WebCore::SharedBuffer &gt; &gt;::encode): Deleted.
(IPC::ArgumentCoder&lt;RefPtr&lt;WebCore::SharedBuffer &gt; &gt;::decode): Deleted.
(IPC::ArgumentCoder&lt;Ref&lt;WebCore::SharedBuffer &gt; &gt;::encode): Deleted.
(IPC::ArgumentCoder&lt;Ref&lt;WebCore::SharedBuffer &gt; &gt;::decode): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.h:

Canonical link: <a href="https://commits.webkit.org/251492@main">https://commits.webkit.org/251492@main</a>
git-svn-id: <a href="https://svn.webkit.org/repository/webkit/trunk@295487">https://svn.webkit.org/repository/webkit/trunk@295487</a> 268f45cc-cd09-0410-ab3c-d52691b4dbfc
</pre>

### CODE REVIEW

File: Source/WebKit/Shared/WebCoreArgumentCoders.cpp [hunk 1]
```diff
@@ -168,91 +168,13 @@ DEFINE_SIMPLE_ARGUMENT_CODER_FOR_SOURCE(DisplayList::SetInlineStrokeColor)
...
-#endif
-
-    return true;
-}
-
 static void encodeTypesAndData(Encoder& encoder, const Vector<String>& types, const Vector<RefPtr<SharedBuffer>>& data)
 {
     ASSERT(types.size() == data.size());
     encoder << types;
     encoder << static_cast<uint64_t>(data.size());
```
Comment (kkinnunen-apple) [inline Source/WebKit/Shared/WebCoreArgumentCoders.cpp]: Same problem here

encoder << data

**Review by kkinnunen-apple — COMMENTED** [review]
Still the redundant functions exist

File: Source/WebKit/Shared/WebCoreArgumentCoders.cpp [hunk 2]
```diff
@@ -1602,10 +1493,10 @@ bool ArgumentCoder<PasteboardCustomData::Entry>::decode(Decoder& decoder, Pasteb
         return false;
```
Comment (kkinnunen-apple) [inline Source/WebKit/Shared/WebCoreArgumentCoders.cpp]: Just a suggestion, not needed to land:
so again, these hasStrings do not conform to the rule: just encode and decode the members as is.

So you can replace all these with something in the spirit of

```
auto platformData = decoder.decode<decltype(PasteboardCustomData::Entry::platformData)>();
...
return PasteboardCustomData::Entry { ... WTFMove(*platformData) };
```

File: Source/WebKit/Shared/WebCoreArgumentCoders.cpp [hunk 3]
```diff
@@ -1586,7 +1477,7 @@ void ArgumentCoder<PasteboardCustomData::Entry>::encode(Encoder& encoder, const
     bool hasBuffer = std::holds_alternative<Ref<SharedBuffer>>(platformData);
     encoder << hasBuffer;
     if (hasBuffer)
-        encodeSharedBuffer(encoder, std::get<Ref<SharedBuffer>>(platformData).ptr());
```
Comment (kkinnunen-apple) [inline Source/WebKit/Shared/WebCoreArgumentCoders.cpp]: (not needed to land, but..)

`encoder << data.type << data.customData << data.platformData;`

File: Source/WebKit/Shared/WebCoreArgumentCoders.cpp [hunk 4]
```diff
@@ -2290,74 +2189,75 @@ void ArgumentCoder<FilterOperation>::encode(Encoder& encoder, const FilterOperat
     ASSERT_NOT_REACHED();
 }
```
Comment (kkinnunen-apple) [inline Source/WebKit/Shared/WebCoreArgumentCoders.cpp]: (nothing to be done for landing, but..)
If I understand correctly, the filter operations changes don't have much to do with shared buffer modifications? (eg. these are drive-by changes?)

**Review by kkinnunen-apple — APPROVED** [review]
nice!

File: Source/WebKit/Shared/WebCoreArgumentCoders.cpp [hunk 5]
```diff
@@ -420,11 +312,10 @@ std::optional<DOMCacheEngine::Record> ArgumentCoder<DOMCacheEngine::Record>::dec
         return std::nullopt;
 
     if (hasSharedBufferBody) {
-        RefPtr<SharedBuffer> buffer;
-        if (!decodeSharedBuffer(decoder, buffer))
+        if (auto buffer = decoder.decode<RefPtr<SharedBuffer>>())
+            responseBody = buffer->releaseNonNull();
```
Comment (kkinnunen-apple) [inline Source/WebKit/Shared/WebCoreArgumentCoders.cpp]: .. and here you are decoding a refptr
(source of your bug that causes the asserts in tests)

This is the reason why it would be better to do the encoder in consistent manner, see above.

File: Source/WebKit/Shared/WebCoreArgumentCoders.cpp [hunk 6]
```diff
@@ -364,7 +256,7 @@ void ArgumentCoder<DOMCacheEngine::Record>::encode(Encoder& encoder, const DOMCa
 
     WTF::switchOn(record.responseBody, [&](const Ref<SharedBuffer>& buffer) {
         encoder << true;
-        encodeSharedBuffer(encoder, buffer.ptr());
+        encoder << buffer;
```
Comment (kkinnunen-apple) [inline Source/WebKit/Shared/WebCoreArgumentCoders.cpp:259]: here you are encoding a ref...

File: Source/WebKit/Shared/WebCoreArgumentCoders.cpp [hunk 7]
```diff
@@ -364,7 +256,7 @@ void ArgumentCoder<DOMCacheEngine::Record>::encode(Encoder& encoder, const DOMCa
 
     WTF::switchOn(record.responseBody, [&](const Ref<SharedBuffer>& buffer) {
```
Comment (kkinnunen-apple) [inline Source/WebKit/Shared/WebCoreArgumentCoders.cpp:257]: so the consistent manner:
"all encoders should do is just encoder << .... << .. <<. ..;"

so here you could do just
`encoder << record.responseBody;
`

**Review by kkinnunen-apple — APPROVED** [review]

File: Source/WebKit/Shared/WebCoreArgumentCoders.cpp (hunk 7 above)
Comment (jyavenard) [reply Source/WebKit/Shared/WebCoreArgumentCoders.cpp:257]: I'll take those in the follow-up patch

File: Source/WebKit/Shared/WebCoreArgumentCoders.cpp (hunk 2 above)
Comment (jyavenard) [reply Source/WebKit/Shared/WebCoreArgumentCoders.cpp]: nice

File: Source/WebKit/Shared/WebCoreArgumentCoders.cpp (hunk 6 above)
Comment (jyavenard) [reply Source/WebKit/Shared/WebCoreArgumentCoders.cpp:259]: thanks for spotting that.

That's another reason on why I think that having the encoder/decoder for Ref<T> use the RefPtr<T> so that encoder/decoder could be used interchangeably.

**Comment by jyavenard** [comment]
thanks for the review

File: Source/WebKit/Shared/WebCoreArgumentCoders.cpp (hunk 7 above)
Comment (jyavenard) [reply Source/WebKit/Shared/WebCoreArgumentCoders.cpp:257]: I'm puzzled with those on how to perform the decoding side.
   ` auto responseBody = decoder.decode<WebCore::DOMCacheEngine::ResponseBody>();`

gives:
```
In file included from /Users/jyavenard/Work/webkit/OpenSource/WebKitBuild/Debug/DerivedSources/WebKit/unified-sources/UnifiedSource41.cpp:7:
/Users/jyavenard/Work/webkit/OpenSource/Source/WebKit/Shared/WebCoreArgumentCoders.cpp:298:33: note: in instantiation of function template specialization 'IPC::Decoder::decode<std::variant<std::nullptr_t, WTF::Ref<WebCore::FormData>, WTF::Ref<WebCore::SharedBuffer>>>' requested here
    auto responseBody = decoder.decode<WebCore::DOMCacheEngine::ResponseBody>();
                                ^
In file included from /Users/jyavenard/Work/webkit/OpenSource/WebKitBuild/Debug/DerivedSources/WebKit/unified-sources/UnifiedSource41.cpp:1:
In file included from /Users/jyavenard/Work/webkit/OpenSource/Source/WebKit/Shared/WebCompiledContentRuleList.cpp:27:
In file included from /Users/jyavenard/Work/webkit/OpenSource/Source/WebKit/Shared/WebCompiledContentRuleList.h:30:
In file included from /Users/jyavenard/Work/webkit/OpenSource/Source/WebKit/Shared/WebCompiledContentRuleListData.h:30:
In file included from /Users/jyavenard/Work/webkit/OpenSource/Source/WebKit/Platform/SharedMemory.h:31:
In file included from /Users/jyavenard/Work/webkit/OpenSource/WebKitBuild/Debug/usr/local/include/wtf/ThreadSafeRefCounted.h:30:
In file included from /Users/jyavenard/Work/webkit/OpenSource/WebKitBuild/Debug/usr/local/include/wtf/MainThread.h:35:
In file included from /Users/jyavenard/Work/webkit/OpenSource/WebKitBuild/Debug/usr/local/include/wtf/ThreadingPrimitives.h:37:
In file included from /Users/jyavenard/Work/webkit/OpenSource/WebKitBuild/Debug/usr/local/include/wtf/WallTime.h:29:
In file included from /Users/jyavenard/Work/webkit/OpenSource/WebKitBuild/Debug/usr/local/include/wtf/GenericTimeMixin.h:28:
In file included from /Users/jyavenard/Work/webkit/OpenSource/WebKitBuild/Debug/usr/local/include/wtf/Seconds.h:28:
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.0.Internal.sdk/usr/include/c++/v1/optional:694:41: note: candidate constructor not viable: no known conversion from 'std::optional<FormData>' to 'const std::optional<WTF::Ref<WebCore::FormData>> &' for 1st argument
    _LIBCPP_INLINE_VISIBILITY constexpr optional(const optional&) = default;
                                        ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.0.Internal.sdk/usr/include/c++/v1/optional:695:41: note: candidate constructor not viable: no known conversion from 'std::optional<FormData>' to 'std::optional<WTF::Ref<WebCore::FormData>> &&' for 1st argument
    _LIBCPP_INLINE_VISIBILITY constexpr optional(optional&&) = default;
                                        ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.0.Internal.sdk/usr/include/c++/v1/optional:696:41: note: candidate constructor not viable: no known conversion from 'std::optional<FormData>' to 'std::nullopt_t' for 1st argument
    _LIBCPP_INLINE_VISIBILITY constexpr optional(nullopt_t) noexcept {}
                                        ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.0.Internal.sdk/usr/include/c++/v1/optional:720:15: note: candidate template ignored: substitution failure [with _Up = std::optional<WebCore::FormData>]: no member named '_EnableIfImpl' in 'std::_MetaBase<false>'
    constexpr optional(_Up&& __v)
              ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.0.Internal.sdk/usr/include/c++/v1/optional:735:35: note: candidate template ignored: substitution failure [with _Up = WebCore::FormData]: no member named '_EnableIfImpl' in 'std::_MetaBase<false>'
    _LIBCPP_CONSTEXPR_AFTER_CXX17 optional(const optional<_Up>& __v)
                                  ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.0.Internal.sdk/usr/include/c++/v1/optional:753:35: note: candidate template ignored: substitution failure [with _Up = WebCore::FormData]: no member named '_EnableIfImpl' in 'std::_MetaBase<false>'
    _LIBCPP_CONSTEXPR_AFTER_CXX17 optional(optional<_Up>&& __v)
                                  ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.0.Internal.sdk/usr/include/c++/v1/optional:706:24: note: explicit constructor is not a candidate
    constexpr explicit optional(_InPlaceT, _Args&&... __args)
                       ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.0.Internal.sdk/usr/include/c++/v1/optional:727:24: note: explicit constructor is not a candidate
    constexpr explicit optional(_Up&& __v)
                       ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.0.Internal.sdk/usr/include/c++/v1/optional:743:44: note: explicit constructor is not a candidate
    _LIBCPP_CONSTEXPR_AFTER_CXX17 explicit optional(const optional<_Up>& __v)
                                           ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.0.Internal.sdk/usr/include/c++/v1/optional:761:44: note: explicit constructor is not a candidate
    _LIBCPP_CONSTEXPR_AFTER_CXX17 explicit optional(optional<_Up>&& __v)
                                           ^
In file included from /Users/jyavenard/Work/webkit/OpenSource/WebKitBuild/Debug/DerivedSources/WebKit/unified-sources/UnifiedSource41.cpp:2:
In file included from /Users/jyavenard/Work/webkit/OpenSource/Source/WebKit/Shared/WebCompiledContentRuleListData.cpp:31:
In file included from /Users/jyavenard/Work/webkit/OpenSource/Source/WebKit/Platform/IPC/ArgumentCoders.h:28:
/Users/jyavenard/Work/webkit/OpenSource/Source/WebKit/Platform/IPC/ArgumentCoder.h:63:20: error: no viable conversion from returned value of type 'RefPtr<WebCore::FormData>' to function return type 'std::optional<FormData>'
            return T::decode(decoder);
                   ^~~~~~~~~~~~~~~~~~
```
